Show file and line information in json output

brianmoose opened the pull request:
This stores file and line information in `Factory`, `ActionFactory`, and `Syntax` by modifying the various register macros.
To get the info for `Syntax` I added a couple of macros that can be used instead of the `syntax.registerActionSyntax()`.
I made it so that this is not required so that existing code should still work.
This is mainly for #7855 . @aeslaughter Does this work for you?

Commit:
Change register macros to add __FILE__ and __LINE__
Add methods to get file/line information for objects and syntax.

refs #7855

Commit:
Output file/line info for actions and syntax.

Commit:
Use new registerSyntax and registerSyntaxTask macros

aeslaughter commented on framework/src/parser/Parser.C:
```diff
@@ -530,17 +530,35 @@ Parser::buildJsonSyntaxTree(JsonSyntaxTree & root) const
+    const std::string act_name = act_names.first;
+    InputParameters action_obj_params = _action_factory.getValidParams(action);
+    // all_names.push_back(std::make_tuple(iter.first, act_info,
+    // _syntax.getLineInfo(act_info._action, act_info._task)));
```
Do you need these?

aeslaughter reviewed (commented):
This should probably include a test of some sort.

aeslaughter commented:
@permcody You should probably look at this before it is merged, it is messing with registration. I will also probably pull this down and use it a bit to see how it will work with MooseDocs.

Commit:
Add FileLineInfo and FileLineInfoMap classes for easier mapping of keys
to line information.
Add test for file line information.
Remove commented out bad code.

permcody commented on framework/include/actions/ActionFactory.h:
```diff
@@ -79,8 +86,17 @@ class ActionFactory
     _name_to_build_info.insert(std::make_pair(name, build_info));
     _task_to_action_map.insert(std::make_pair(task, name));
+    if (!file.empty() && line >= 0)
+      _name_to_line.insert(std::make_pair(std::make_pair(name, task), std::make_pair(file, line)));
```
Holy Dooley!

permcody commented on framework/include/actions/ActionFactory.h:
```diff
@@ -115,6 +131,7 @@ class ActionFactory
 
   std::multimap<std::string, BuildInfo> _name_to_build_info;
 
+  std::map<std::pair<std::string, std::string>, std::pair<std::string, int>> _name_to_line;
```
At what point do you finally decide to build a small struct? 😆

permcody commented on framework/src/parser/Parser.C:
```diff
@@ -530,17 +530,33 @@ Parser::buildJsonSyntaxTree(JsonSyntaxTree & root) const
+    const auto & act_info = act_names.second;
+    const std::string & action = act_info._action;
+    const std::string & task = act_info._task;
+    const std::string act_name = act_names.first;
```
Yeah... Starting to think a little struct would be cleaner.

brianmoose replied:
Agreed.

permcody commented on framework/src/parser/Parser.C:
```diff
@@ -530,17 +530,33 @@ Parser::buildJsonSyntaxTree(JsonSyntaxTree & root) const
+    if (params_added)
+    {
+      auto tasks = _action_factory.getTasksByAction(action);
+      for (auto && t : tasks)
```
Universal reference: We looked at this when we first transitioned and decided we didn't like them for some reason. We've been just using normal references and marking the type as const when possible. I may have to revisit this and rethink why we didn't like them.

permcody commented on framework/src/parser/Syntax.C:
```diff
@@ -228,3 +238,17 @@ Syntax::verifyMooseObjectTask(const std::string & base, const std::string & task
+  if (iters.first != iters.second)
+    for (auto && it = iters.first; it != iters.second; ++it)
+      if (std::get<0>(it->second) == action && std::get<1>(it->second) == task)
+        return std::make_pair(std::get<2>(it->second), std::get<3>(it->second));
```
This syntax is new for me. Are you digging out part of the second pair here?

brianmoose replied:
In `Syntax` we store a tuple, so these are just getting access to each value.
I think another structure would be good here too, especially given how this map is currently accessed.

permcody commented on framework/src/utils/JsonSyntaxTree.C:
```diff
@@ -120,22 +124,30 @@ JsonSyntaxTree::addParameters(const std::string & parent, const std::string & pa
-    json[action]["tasks"].append(task_name);
+    json[action]["description"] = params->getClassDescription();
     json[action]["action_path"] = path;
+    auto & fi = json[action]["file_info"];
```
Bracket access on RHS: This is our infamous "accidental insertion" pattern for std::map. Is this access OK here if the json object is empty? If not can you add a MOOSE assert or something?

brianmoose replied:
Yes, this and the other one are intentional. This always puts a `file_info` in the json, it just might be `null` if there is no information. This allows the parser to not have to check if the key exists. But I can remove it if you like.

permcody commented on framework/src/utils/JsonSyntaxTree.C:
```diff
@@ -197,3 +209,15 @@ JsonSyntaxTree::buildOutputString(
+                              const std::pair<std::string, int> & lineinfo)
+{
+  moosecontrib::Json::Value & json = getJson("", path, false);
+  auto & fi = json[action]["tasks"][task_name]["file_info"];
```
Again more brackets on RHS: Please verify that this is always OK.

permcody reviewed (changes requested):
This looks OK to me. I think the use of maps with pairs as both keys and values is a little over-clever though. I'd prefer a slightly cleaner design.

brianmoose commented:
@aeslaughter So I added a test and it revealed a difference in how gcc and clang process `__FILE__` in c++. Clang seems to place the full path name there while gcc places the filename as specified on the command line (ie just `src/parser/Parser.C`). Is that going to be a problem?

brianmoose commented:
Upon further investigation this doesn't seem to be a gcc vs clang thing. Probably a mac vs linux thing? Anyway, on mac with clang we get the full source file path. On linux, with gcc or clang, we get the path relative to where `make` was called.

aeslaughter commented:
I can deal with the relative path stuff.

aeslaughter commented on framework/include/utils/FileLineInfo.h:
```diff
@@ -0,0 +1,120 @@
+   * @param file file
+   * @param line line number
+   */
+  void addInfo(const std::string & key0, const std::string & file, int line);
```
Is it possible to use C++11 variadic arguments for the keys? Just to avoid having three definitions for the add and get methods.

permcody replied:
Even if that worked, it's probably not the best solution. You could use a tuple, or small struct instead though. However, I don't think having several overloads is bad. They have the same semantic meaning and the implementations are clean. Trying to condense them will likely make the code more difficult to understand.

aeslaughter commented on framework/src/parser/Parser.C:
```diff
@@ -530,17 +530,33 @@ Parser::buildJsonSyntaxTree(JsonSyntaxTree & root) const
+                                           action,
+                                           true,
+                                           &action_obj_params,
+                                           _syntax.getLineInfo(act_name, action, ""));
```
Should the addParameters just accept "act_info" from which it can do the extraction of the various parts?

permcody replied:
Probably not, That would be spreading encapsulation from the Parser into InputParameters which is unrelated. This is a better design even if the call looks messier.

brianmoose replied:
Also, there is another call to addParameters below this that doesn't use information from act_info.

Commit:
Fix style and error in conflict resolution.

aeslaughter commented:
@permcody If you are happy with this, then you can merge away.
